src/pint/gridutils.py

```diff
@@ -1,16 +1,278 @@
 
 import astropy.constants as const
 import astropy.units as u
 import numpy as np
 
 import pint.utils
 
-__all__ = ["grid_chisq", "grid_chisq_mp", "plot_grid_chisq"]
+log = logging.getLogger(__name__)
+
+__all__ = ["doonefit", "grid_chisq", "grid_chisq_derived", "plot_grid_chisq"]
+
+
+def doonefit(ftr, parnames, parvalues):
+    """Worker process that computes one fit with specified parameters fixed
+
+    Parameters
+    ----------
+    ftr : pint.fitter.Fitter
+    parnames : list
+        Names of the parameters to grid over
+    parvalues : list
+        List of parameter values to grid over (each should be 1D array of astropy.units.Quantity)
+    """
+    # Make a full copy of the fitter to work with
+    myftr = copy.deepcopy(ftr)
+    for parname, parvalue in zip(parnames, parvalues):
+        # Freeze the  params we are going to grid over and set their values
+        # All other unfrozen parameters will be fitted for at each grid point
+        getattr(myftr.model, parname).frozen = True
+        getattr(myftr.model, parname).quantity = parvalue
+    myftr.fit_toas()
+    return myftr.resids.chi2
+
+
+def grid_chisq(
+    ftr,
+    parnames,
+    parvalues,
+    ncpu=None,
+    executor=concurrent.futures.ProcessPoolExecutor,
+    chunksize=1,
+    printprogress=True,
+):
+    """Compute chisq over a grid of parameters
+
+    Parameters
+    ----------
+    ftr : pint.fitter.Fitter
+        The base fitter to use.
+    parnames : list
+        Names of the parameters to grid over
+    parvalues : list
+        List of parameter values to grid over (each should be 1D array of astropy.units.Quantity)
+    ncpu : int, optional
+        Number of processes to use in parallel. Default is number of CPUs available.
+        If `ncpu` is 1, then use single-threaded version
+    executor : concurrent.futures.Executor
+        Executor object to run multiple processes in parallel (if `ncpu` is None or >1)
+    chunksize : int
+        Size of the chunks for :class:`concurrent.futures.ProcessPoolExecutor` parallel execution.
+        Ignored for :class:`concurrent.futures.ThreadPoolExecutor`
+    printprogress : bool, optional
+        Print indications of progress for single processor only
+
+    Returns
+    -------
+    np.ndarray : array of chisq values
+    """
+    if ncpu is None or ncpu > 1:
+        if ncpu is None:
+            # Use al available CPUs
+            ncpu = multiprocessing.cpu_count()
+
+    # Save the current model so we can tweak it for gridding, then restore it at the end
+    savemod = ftr.model
+    gridmod = copy.deepcopy(ftr.model)
+    ftr.model = gridmod
+
+    # Freeze the  params we are going to grid over
+    for parname in parnames:
+        getattr(ftr.model, parname).frozen = True
+
+    # All other unfrozen parameters will be fitted for at each grid point
+    out = np.meshgrid(*parvalues)
+    chi2 = np.zeros(out[0].shape)
+    if ncpu > 1:
+        with executor(max_workers=ncpu) as e:
```

Comment:
I would suggest accepting an already-created executor as an argument - executors might require various startup arguments that this function doesn't know about. So if you want to run this with a multiprocessing executor, you just call this function in a with statement and pass it the executor. This also helps keep executors around for multiple calls, if they're expensive to create (ssh to a bunch of machines...)

Reply:
I thought about that.  I think for more complicated cases that would definitely be better, but there might also be simpler cases where creating it is better.  I'll see if I can add a little flexibility to do both.

Reply:
Maybe I should be clearer: using an executor to run parallel tasks has a well-defined interface, but creating executors does not.

Reply:
OK, I didn't realize that.  I still think there should be a way to run this with no external work, so I'll see what looks reasonable.